packages/app/src/client/views/tenants/TenantPicker.tsx

```diff
@@ -43,17 +44,27 @@ const TenantPicker = () => {
   const dispatch = useDispatch();
   const history = useHistory();
   const tenants = useTenantList();
+  const currentTenant = useSelectedTenant();
 
   const { activatePickerWithText } = usePickerService(
     {
       activationPrefix: "tenant:",
       options: tenants ? tenants.map(tenantToPickerOption) : [],
       onSelected: option => {
         dispatch(selectedTenantChanged({ tenant: option.data }));
-        history.push(`/tenant/${option.id}`); // todo: would be nice to keep the current path and simply change the tenant rather than send user back to the home for the tenant
+        const { pathname } = history.location;
+        const newRoute = currentTenant
+          // if we're on a page with a tenant selected, stay on page and just replace the tenant
+          ? pathname.replace(
+              `/tenant/${currentTenant.name}`,
+              `/tenant/${option.id}`
+            )
+          // if no tenant is currently selected, reroute to tenant root route
```

Comment:
this is currently the case in routes for cluster admin

Reply:
perhaps we can memoize the last value of `currentTenant.name` where the routes are declared and then we update that "root route" as referenced in the comment, to route to `/tenant/<memoized_tenant>` instead of `/tenant/system`?

Reply:
and only fallback to "system" tenant if the value of the memoized tenant is nullish

Reply:
not sure if memoization is the right tool for this job. React even goes as far as to discourage using `useMemo` for these cases.

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance. - [React Docs](https://reactjs.org/docs/hooks-reference.html#usememo)

For now I just added a new hook for this use case (that includes tests) [here](https://github.com/opstrace/opstrace/pull/933/commits/d3ef3d2231df2e36bc563aec6da9409c0f4316ba).